core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -293,13 +292,16 @@ private[spark] class TaskSchedulerImpl(
     // Also track if new executor is added
     var newExecAvail = false
     for (o <- offers) {
-      executorIdToHost(o.executorId) = o.host
-      executorIdToTaskCount.getOrElseUpdate(o.executorId, 0)
       if (!executorsByHost.contains(o.host)) {
         executorsByHost(o.host) = new HashSet[String]()
+      }
+      if (!executorIdToHost.contains(o.executorId)) {
```

Comment:
At first I was keeping them separate, but I discovered that https://github.com/apache/spark/pull/13603 was flaky without the changes proposed here.  So now that PR just has all of this one merged into it.  My intention is that we merge this one first, including the commit which does the test cleanup.  (I don't think it particularly matters which PR the test cleanup goes in, but might as well put it in the first one.)